api/jobset/v1alpha2/jobset_types.go

```diff
@@ -134,6 +136,10 @@ type JobSetStatus struct {
 	// RestartsCountTowardsMax tracks the number of times the JobSet has restarted that counts towards the maximum allowed number of restarts.
 	RestartsCountTowardsMax int32 `json:"restartsCountTowardsMax,omitempty"`
 
+	// Phase of the JobSet.
+	// +kubebuilder:default="Running"
+	Phase string `json:"phase,omitempty"`
```

Comment:
@ahg-g @danielvegamyhre @kannon92 Do we really need to introduce the phase field? Having the separate state machine opposed to conditions will bring us high maintenance costs, and they potential cause bugs.

Actually, this approach is not the recommended approach described in the API documentation. 
So, I would recommend adding new condition types instead of a new phase field.

> Some resources in the v1 API contain fields called phase, and associated message, reason, and other status fields. The pattern of using phase is deprecated. Newer API types should use conditions instead. Phase was essentially a state-machine enumeration field, that contradicted [system-design principles](https://git.k8s.io/design-proposals-archive/architecture/principles.md#control-logic) and hampered evolution, since [adding new enum values breaks backward compatibility](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md). Rather than encouraging clients to infer implicit properties from phases, we prefer to explicitly expose the individual conditions that clients need to monitor. Conditions also have the benefit that it is possible to create some conditions with uniform meaning across all resource types, while still exposing others that are unique to specific resource types. See [#7856](http://issues.k8s.io/7856) for more details and discussion.

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties

Reply:
Although, at a glance, the phase field is a simpler solution, I don't think so.
Because the phase field will bring us needs to maintain 2 dimensions state machine between conditions and phase, I think that maintaining 2 dimensions state machine is not easy.

Reply:
The phase should be a direct translation of conditions, it shouldn't represent a different state. I suggest we limit it to the terminal state of the jobset, which I think is what the original ask was about. TerminalState: success|failed, if more details are needed, then one needs to look that up in the conditions.

Reply:
> The phase should be a direct translation of conditions, it shouldn't represent a different state. I suggest we limit it to the terminal state of the jobset, which I think is what the original ask was about. TerminalState: success|failed, if more details are needed, then one needs to look that up in the conditions.

+1, I think these terminal states will be easy to implement and maintain. The "running" phase would be much more ambiguous / hard to define and maintain.

Reply:
OK, I will collect the comments and make changes. 
If I change it to the TerminalState field, will there still be any problems according to your suggestion? @tenzen-y